controllers/teleport_controller/teleport_controller.cpp

```diff
@@ -0,0 +1,138 @@
+        std::cout << "X: " << target_translation_vec[0];
+        std::cout << " Y: " << target_translation_vec[1];
+        std::cout << " Z: " << target_translation_vec[2] << std::endl;
+
+        // Prepare new location
+
+        /* 0, 0, 0 is centre of the playing field.
+         *     X ranges from -5.4 - 5.4, the positive value on the left hand side when looking
+         *     from red
+         *     Y ranges from -8 - 8, the positive value on the red goal side
+         *     Z should be set to 3.2 to be on the ground
+         *     If the robot teleports into an existing object it may run into issues for that
+         *     image only, after resetPhysics is should return to a regular state
+         */
+
+        double newPos[3];
+        newPos[0] = 5.4 - (rand() % 1080) / 100;  // target_translation_vec[0] + 1;
+        newPos[1] = 8 - (rand() % 1600) / 100;    // target_translation_vec[1] + 0;
+        newPos[2] = 0.32;                         // target_translation_vec[2] + 0;
+
+        // Set new location
+        target_translation_field->setSFVec3f(newPos);
+
+        // Grab the current rotation field of the robot to modify
+        Field* target_rotation_field = target->getField("rotation");
+        // Convert the field to a vector to output to console
+        const double* target_rotation_vec = target_rotation_field->getSFRotation();
+
+        // Loop 4 times, one for each direction to face
+        for (int i = 0; i < rotationsSize; i++) {
+            // Output current rotation
+            std::cout << "Rotation: " << std::endl;
+            std::cout << "X: " << target_rotation_vec[0];
+            std::cout << " Y: " << target_rotation_vec[1];
+            std::cout << " Z: " << target_rotation_vec[2];
+            std::cout << " α: " << target_rotation_vec[3] << std::endl;
+
+            /* Prepare new rotation. These are saved in rotations matrix as the axis-angle
+               calculation is rough to calculate on the fly*/
+
+            double newRot[4];
+            newRot[0] = rotations[i][0];
+            newRot[1] = rotations[i][1];
+            newRot[2] = rotations[i][2];
+            newRot[3] = rotations[i][3];
+
+            // Set new rotations
+            target_rotation_field->setSFRotation(newRot);
+
+            // This line is very important. The robot blows itself up after a 1:05 minutes otherwise
+            // Teleporting is strenuous work apparently
+            target->resetPhysics();
+
+            /* Make robot stand up and look around. This preset is slow and I'm not super
+             *     happy with it but it does the job for now. Below is a work in progress to
+             *     manually move the robot motors to where we want them
+             */
+            mMotionManager->playPage(54);
+            // Other presets, comment out as desired
+            mMotionManager->playPage(15);
+            mMotionManager->playPage(1);
+        }
+    };
+
+    delete supervisor;
```

Comment:
raw pointers bad. RAII is based

Reply:
Ahaha again just a copypasta from the sample code, it worked so I kept it but can change it

Reply:
It's not that the supervisor object shouldn't be deleted, it's that smart pointers do the deletion for us when they go out of scope. Calling delete explicitly means you've used raw pointers (basically any time you use the `new` keyword, it's a raw pointer), and raw pointers are just a fast track to memory leaks and other fun issues.

Reply:
Have a look in the networking PR #1 for how I used a `std::unique_ptr` (a type of smart pointer) in that controller.

Reply:
Yeah alright I'll have a look

Reply:
Just a quick mention about one of the easiest reasons to see when a raw pointer will do a memory leak

```
int* pointer = new int;
if(i_feel_like_it) {
    throw std::runtime_error();
}
delete pointer;
```

In this case you have a memory leak because you never delete the pointer (you throw out of the function first)
If it were a smart pointer, it would be deleted

Reply:
Yeah okay that makes heaps of sense, cool, I'll probably just change it around so the whole thing runs on the stack lol, it's fairly lightweight and that way I don't need to worry about it at all